tools/java/org/unicode/cldr/api/CldrDataSupplier.java

```diff
@@ -0,0 +1,241 @@
+package org.unicode.cldr.api;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.LinkedHashMultimap;
+import com.google.common.collect.Multimap;
+import org.unicode.cldr.api.CldrData.PrefixVisitor;
+import org.unicode.cldr.api.CldrData.ValueVisitor;
+import org.unicode.cldr.util.CLDRFile;
+import org.unicode.cldr.util.Factory;
+import org.unicode.cldr.util.SimpleFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static org.unicode.cldr.api.CldrDataType.LDML;
+
+/**
+ * The main API for accessing {@link CldrPath} and {@link CldrValue} instances for CLDR data. This
+ * API abstracts the data sources, file names and other implementation details of CLDR to provide
+ * a clean way to access CLDR data.
+ *
+ * <p>{@code CldrData} instances are obtained from an appropriate {@code CldrDataSupplier}, and
+ * accept a {@link ValueVisitor} or {@link PrefixVisitor} to iterate over the data.
+ *
+ * <p>For example the following code prints every value (including its associated distinguishing
+ * path) in the BCP-47 data in DTD order:
+ * <pre>{@code
+ *   CldrDataSupplier supplier = CldrDataSupplier.forFilesIn(rootDir);
+ *   CldrData bcp47Data = supplier.getDataForType(CldrDataType.BCP47);
+ *   bcp47Data.accept(PathOrder.DTD, System.out::println);
+ * }</pre>
+ *
+ * <p>Note that while the paths of values visited in a single {@link CldrData} instance are unique,
+ * there is nothing to prevent duplication between multiple data sources. This is particularly
+ * important when considering "ordered" elements with a sort index, since it represents "encounter
+ * order" and so any merging of values would have to track and rewrite sort indices carefully. It
+ * is recommended that if multiple {@code CldrData} instances are to be processed, users ensure
+ * that no path prefixes be shared between them. See also {@link CldrPath#getSortIndex()}.
+ *
+ * <p>Note that because the distinguishing paths associated with a {@link CldrValue} are unique per
+ * visitation, the special "version" path/value must be omitted (e.g. "//ldml/version") since it
+ * would otherwise appear multiple times. This should be fine, since the version is always available
+ * via {@link #getCldrVersionString()} and this mechanism is scheduled for deprecation anyway.
+ */
+public abstract class CldrDataSupplier {
+    /** Returns the current CLDR version string (e.g. {@code "36"}). */
+    public static String getCldrVersionString() {
+        return CLDRFile.GEN_VERSION;
```

Comment:
i assume this is here for API locality, a comment to that effect might be good so that it's not refactored out in the future.

Reply:
Yes, there's a bunch of little things like this just so the api package is all that's ever needed.
I'm hoping that we could have a check for not including CLDR APIs in the ICU code (since it's otherwise so trivial to accidentally import things from outside). One way would be to compile against an "empty" JAR that only contains stubs for the API.

Reply:
https://unicode-org.atlassian.net/browse/CLDR-13162